### PR TITLE
CE-94: Collecting PID errors, rather than reporting them one at a time

### PR DESCRIPTION
**Pull request by ialarmedalien**
Also entabbed some whitespace and updated the name of the class in which all this info will be held.

### CODE REVIEW

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 1]
```diff
@@ -155,10 +155,10 @@ public String getPid() {
 	 *                                  relationship type name related to the input string.
 	 */
 	public static RelationshipType getRelationshipType(final String input) {
-		final String lowercaseInput = Util.checkString(input, "input").toLowerCase();
-		if (STRING_TO_REL_TYPE_MAP.containsKey(lowercaseInput)) {
-			return STRING_TO_REL_TYPE_MAP.get(lowercaseInput);
+		final String lowercaseInput = Util.checkString(input, "relationshipType").toLowerCase();
+		if (!STRING_TO_REL_TYPE_MAP.containsKey(lowercaseInput)) {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:159]: switched this so it matches the other enums.

File: src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java [hunk 2]
```diff
@@ -48,33 +48,33 @@ private void assertTitleStringAndOptionalNulls(final PermanentID pid, final Stri
 
 	@Test
 	public void buildMinimal() throws Exception {
-		for (Map.Entry<String, String> mapElement : VALID_PID_MAP.entrySet()) {
-			final PermanentID pid1 = PermanentID.getBuilder(mapElement.getKey()).build();
-			assertTitleStringAndOptionalNulls(pid1, mapElement.getValue());
+		for (Map.Entry<String, String> entry : VALID_PID_MAP.entrySet()) {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java:51]: renamed the `entrySet` variables for consistency between classes

File: src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java [hunk 3]
```diff
@@ -108,96 +108,106 @@ public void buildMaximalWithDifferentRelTypeStrings() throws Exception {
...
+			final PermanentID pid1 = PermanentID.getBuilder(entry.getKey())
 					.withDescription(DESC_STRING).withDescription(null)
 					.withRelationshipType(REL_TYPE).withRelationshipType((RelationshipType) null)
 					.build();
-			assertTitleStringAndOptionalNulls(pid1, mapElement.getValue());
+			assertTitleStringAndOptionalNulls(pid1, entry.getValue());
+		}
+	}
+
+	private void buildPermanentIDFailWithError(final PermanentID.Builder builder, final String error) {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java:165]: refactor to move failure checks to a separate function

File: src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java [hunk 4]
```diff
@@ -108,96 +108,106 @@ public void buildMaximalWithDifferentRelTypeStrings() throws Exception {
...
-		} catch (Exception got) {
-			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
-					"Invalid relationship type: unmutualism"));
+		buildPermanentIDFailWithError(
+				PermanentID.getBuilder("ID:permanent").withRelationshipType("unmutualism"),
+				"Invalid relationshipType: unmutualism");
+	}
+
+	@Test
+	public void buildFailInvalidFields() throws Exception {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/PermanentIDTest.java:203]: new test with collated errors

File: src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java [hunk 5]
```diff
@@ -56,19 +56,28 @@ public void testGetRelationshipType() throws Exception {
 		}
 
 		final String[] testCrossrefInputs = {
-                        "hasmanifestation",
-                        "   \n  has_manifestation   \t\t",
-                        "CROSSREF:HASMANIFESTATION",
-                        "  crossref:has_manifestation  "
-                };
+			"hasmanifestation",
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:59]: entabbing whitespace

File: src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java [hunk 6]
```diff
@@ -56,19 +56,28 @@ public void testGetRelationshipType() throws Exception {
...
+		for (final String testInput : testCrossrefInputs) {
+			assertThat("incorrect role",
+					RelationshipType.getRelationshipType(testInput),
+					is(RelationshipType.HAS_MANIFESTATION));
+		}
 
-        }
+	}
+
+	private void getRelationshipTypeFail(final String input, final String error) {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:73]: extract failure checking function

File: src/us/kbase/workspace/database/provenance/PermanentID.java [hunk 7]
```diff
@@ -42,8 +44,8 @@ public Optional<String> getDescription() {
 	/**
 	 * Gets the relationship between the ID and some other entity. For
 	 * example, when a {@link PermanentID] class is used to represent
-	 * objects in a {@link Resource}'s 'related_identifiers' field, this
-	 * field captures the relationship between the {@link Resource} and
+	 * objects in a {@link CitationMetadata}'s 'related_identifiers' field, this
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/PermanentID.java:47]: Change the name of the class representing the citation metadata to something more accurate

File: src/us/kbase/workspace/database/provenance/PermanentID.java [hunk 8]
```diff
@@ -87,17 +90,23 @@ public static class Builder {
...
+			}
 		}
 
 		/**
 		 * Sets a free text description of the ID.
 		 *
-		 * @param description the description. Null, whitespace, or
-		 *                    the empty string will remove any
-		 *                    current content in the builder.
+		 * @param description
```
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/PermanentID.java:106]: "magic"

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 9]
```diff
@@ -155,10 +155,10 @@ public String getPid() {
 	 *                                  relationship type name related to the input string.
 	 */
 	public static RelationshipType getRelationshipType(final String input) {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:157]:
```suggestion
	public static RelationshipType getRelationshipType(final String relationshipType) {
```
Make error message match the variable name for easier debugging

**Review by MrCreosote — CHANGES_REQUESTED** [review]
1 change, plus standard spiel re maintenance burden etc etc.

File: src/us/kbase/workspace/database/provenance/PermanentID.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/PermanentID.java:106]: 😖

**Review by MrCreosote — APPROVED** [review]
LGTM, just tests need to pass

**Pull request merged**